File: backend/app/api/v1/endpoints/observability.py
# (c) 2026 AgentFlow-Eval
"""Business KPIs, slow tasks, error topology (read-only)."""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Depends, Query, Request
from sqlalchemy.ext.asyncio import AsyncSession

from app.config import settings
from app.core.dependencies import get_db
from app.core.observability.business_kpis import compute_kpis
from app.core.observability.slow_tasks import list_slow_tasks, list_slow_tasks_db
from app.core.observability.tracing import get_trace_id
from app.core.profiles import profile_status
from app.core.rbac import Permission, require_permission

router = APIRouter()


@router.get("/kpis")
@require_permission(Permission.TASK_READ)
async def get_kpis(
    request: Request,
    days: int = Query(7, ge=1, le=90),
    session: AsyncSession = Depends(get_db),
) -> dict[str, Any]:
    """Business KPI dashboard data."""
    if not getattr(settings, "OBSERVABILITY_KPI_ENABLED", True):
        return {"enabled": False, "items": {}}
    actor = getattr(request.state, "actor", None)
    data = await compute_kpis(session, days=days, actor=None)
    return {
        "enabled": True,
        "kpis": data,
        "deploy": profile_status(),
        "viewer": actor,
        "trace_id": getattr(request.state, "request_id", None) or get_trace_id() or None,
    }


@router.get("/slow-tasks")
@require_permission(Permission.TASK_READ)
async def get_slow_tasks(
    request: Request,
    limit: int = Query(50, ge=1, le=200),
    source: str = Query(
        "auto",
        description="auto|db|memory — auto prefers durable DB with memory fallback",
    ),
) -> dict[str, Any]:
    """Recent suite/judge runs exceeding SLOW_TASK_THRESHOLD_SEC."""
    src = (source or "auto").lower().strip()
    if src == "memory":
        items = list_slow_tasks(limit=limit)
        for it in items:
            it.setdefault("source", "memory")
        used = "memory"
    elif src == "db":
        items = await list_slow_tasks_db(limit=limit)
        used = "db"
    else:
        items = await list_slow_tasks_db(limit=limit)
        used = "db" if items and items[0].get("source") == "db" else "memory"
        if not items:
            items = list_slow_tasks(limit=limit)
            for it in items:
                it.setdefault("source", "memory")
            used = "memory" if items else used
    return {
        "items": items,
        "total": len(items),
        "source": used,
        "threshold_sec": float(
            getattr(settings, "SLOW_TASK_THRESHOLD_SEC", 30.0) or 30.0
        ),
        "trace_id": getattr(request.state, "request_id", None) or get_trace_id() or None,
    }


@router.get("/error-topology")
@require_permission(Permission.TASK_READ)
async def get_error_topology(
    request: Request,
    days: int = Query(7, ge=1, le=90),
    session: AsyncSession = Depends(get_db),
) -> dict[str, Any]:
    data = await compute_kpis(session, days=days)
    return {
        "window_days": days,
        "topology": data.get("error_topology") or {},
        "by_status": data.get("by_status") or {},
    }

File: frontend/src/api/endpoints/observability.ts
/* (c) 2026 AgentFlow-Eval */
import { apiClient } from "../client";

export type BusinessKpis = {
  window_days: number;
  since?: string;
  tasks_total: number;
  by_status: Record<string, number>;
  success_rate: number | null;
  avg_metric_score: number | null;
  total_tokens: number;
  avg_trace_latency_ms: number | null;
  error_topology: Record<string, number>;
};

export type KpisResponse = {
  enabled: boolean;
  kpis: BusinessKpis;
  deploy?: Record<string, unknown>;
};

export type SlowTaskItem = {
  stage: string;
  duration_sec: number;
  threshold_sec: number;
  ref_id?: string | null;
  status?: string;
  at?: number;
  trace_id?: string | null;
  extra?: Record<string, unknown>;
};

export const observabilityApi = {
  kpis: (days = 7) =>
    apiClient
      .get<KpisResponse>("/observability/kpis", { params: { days } })
      .then((r) => r.data),
  slowTasks: (limit = 20, source: "auto" | "db" | "memory" = "auto") =>
    apiClient
      .get<{
        items: SlowTaskItem[];
        total: number;
        threshold_sec: number;
        source?: string;
      }>("/observability/slow-tasks", { params: { limit, source } })
      .then((r) => r.data),
  errorTopology: (days = 7) =>
    apiClient
      .get<{ topology: Record<string, number>; by_status: Record<string, number> }>(
        "/observability/error-topology",
        { params: { days } }
      )
      .then((r) => r.data),
};
